### PR TITLE
Add float conversions to and from bytes

### LINKED ISSUE #57492: from_*e_bytes and to_n*_bytes equivalents for floats
These feel appropriate to add for completeness, and would simply chain `from_bits` or `to_bits` with the appropriate integer method.

I'll add a PR soon.

This issue has been assigned to @lzutao via [this comment](https://github.com/rust-lang/rust/issues/57492#issuecomment-508420816).

### PR DESCRIPTION
**Pull request by tbu-**
Use the same API as for integers.

Fixes #57492.

### CODE REVIEW

**Comment by tbu-** [comment]
Will add docs if the general approach is deemed to be okay.

File: src/libcore/num/f32.rs [hunk 1]
```diff
@@ -474,4 +474,65 @@ impl f32 {
         // It turns out the safety issues with sNaN were overblown! Hooray!
         unsafe { mem::transmute(v) }
     }
+
+    #[unstable(feature = "float_to_from_bytes", issue = "0")]
+    #[rustc_const_unstable(feature = "float_to_from_bytes_const")]
+    #[inline]
+    pub const fn to_be_bytes(self) -> [u8; 4] {
+        swap_if_le(self.to_ne_bytes())
```
Comment (scottmcm) [inline src/libcore/num/f32.rs]: We already have `f32::to_bits()` that gives `u32`; I'd expect this to just be `self.to_bits().to_ne_bytes()` (and analogously elsewhere).

**Comment by Dylan-DPC** [comment]
ping from triage @tbu- waiting for your updates on the PR in reply to the review above

**Comment by tbu-** [comment]
I'm not really sure what to put into the examples. Does that work?

**Comment by Dylan-DPC** [comment]
ping from triage @tbu- @aidanhs  any updates on this?

**Comment by tbu-** [comment]
I was wondering whether the examples are fine or what I should extend them with.

**Comment by Centril** [comment]
Ping from triage...

r? @scottmcm 
cc @tbu-

**Comment by scottmcm** [comment]
The code here seems fine, but since it's a new API I think it should have someone from @rust-lang/libs take a look at the APIs and comment if they're interested before going in.

(Not a full FCP, since they're just unstable, but I'm not up-to-date on what they're looking for in new stuff.)

**Comment by SimonSapin** [comment]
I personally feel that the need for these APIs is less pressing than the integer ones, because this operation can already be achieved without unsafe code. (Without `transmute`, in particular.)

But I’m not entirely opposed either. I could see the argument that this more discoverable, it might not be obvious that e.g. `u32::from_be_bytes` and `f32::from_bits` can be combined in this way.

Let’s see what other @rust-lang/libs members think.

**Comment by alexcrichton** [comment]
I'd personally be in favor of being consistent across all the primitive types, so I'd be ok merging and eventually stabilizing this. I don't feel too strongly though and agree @SimonSapin that the pressure isn't so high as there's already methods of doing this, but I'd just fall a little bit on the other side of the fence!

**Comment by alexcrichton** [comment]
@SimonSapin this was pinged back to the libs team recently, do you feel that this *shouldn't* land as unstable to start off with? It looks like there aren't a ton of other opinions about htese APIs.

**Comment by SimonSapin** [comment]
Alright, let’s land this. @tbu-, could you make a few changes?

* Change the tracking issue number to #60446 
* Remove `rustc_const_unstable` attributes as they’re redundant with `unstable` attribute. And when doing FCP to stabilize we’ll do so as `const fn`, no need for a separate FCP just for constness.
* Add some documentation to `from_*` methods and `f64` methods.

**Comment by Centril** [comment]
> * When doing FCP to stabilize we’ll do so as `const fn`, no need for a separate FCP just for constness.

I would suggest keeping `rustc_const_unstable` as it is highly unlikely that you'll get to stabilize the constness concurrently with the usual stabilization. The function `to_bits` isn't `const fn` (and won't be soon) and moreover we do not generally allow floating point related logic in `const fn` because of differences in behavior between runtime and compile time function execution (if there provably cannot be differences in this instance then it may be workable but we need a larger discussion that should not block the initial stabilization).

**Comment by jonas-schievink** [comment]
Ping from triage @tbu-, this still needs a few changes before it can be merged.

**Comment by jonas-schievink** [comment]
> if there provably cannot be differences in this instance then it may be workable

AFAIK, NaN payloads are generally processor-specific, and they can be observed and manipulated using these functions. Perhaps this should not be a `const fn` at all then?

**Comment by Dylan-DPC** [comment]
ping from triage @tbu- any updates?

**Comment by scottmcm** [comment]
@tbu- I'm going to close this for now to keep the backlog clean.  Please do re-open it if you get the chance to address the comments in https://github.com/rust-lang/rust/pull/58756#issuecomment-488385889.

**Comment by lzutao** [comment]
I reworked this in #62366 .

**Pull request closed without merging**